compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1609,44 +1609,66 @@ pub type PlaceholderType = Placeholder<BoundVar>;
 pub type PlaceholderConst = Placeholder<BoundVar>;
 
 /// A `DefId` which is potentially bundled with its corresponding generic parameter
-/// in case `did` is a const argument.
-///
-/// This is used to prevent cycle errors during typeck
-/// as `type_of(const_arg)` depends on `typeck(owning_body)`
-/// which once again requires the type of its generic arguments.
-///
-/// Luckily we only need to deal with const arguments once we
-/// know their corresponding parameters. We (ab)use this by
-/// calling `type_of(param_did)` for these arguments.
+/// in case `did` is a const argument. Used to prevent cycle errors during `typeck`.
 ///
 /// ```rust
 /// #![feature(const_generics)]
 ///
 /// struct A;
 /// impl A {
-///     fn foo<const N: usize>(&self) -> usize { N }
+///     fn foo<const N: usize>(&self) -> [u8; N] { [0; N] }
+///     //           ^ const parameter
 /// }
 /// struct B;
 /// impl B {
-///     fn foo<const N: u8>(&self) -> usize { 42 }
+///     fn foo<const M: u8>(&self) -> usize { 42 }
+///     //           ^ const parameter
 /// }
 ///
 /// fn main() {
 ///     let a = A;
-///     a.foo::<7>();
+///     let _b = a.foo::<7>();
+///     //               ^ const argument
 /// }
 /// ```
+///
+/// Let's look at the call `a.foo::<7>()` here. We do not know which
+/// `foo` is used until we know the type of `a`.
+///
+/// We only know the type of `a` once we are inside of `typeck(main)`.
+/// At that point we also want to know the type of `_b` which depends on
+/// the value of the const argument.
+///
+/// To evaluate that const argument we need to know its type,
+/// which we would get using `type_of(const_arg)`. This requires us to
+/// resolve `foo` as it can be either `usize` or `u8` in this example.
+/// However, resolving `foo` once again requires `typeck(main)` to get the type of `a`,
+/// which results in a cycle.
+///
+/// We only care about the const argument once we know which paramter they instantiate.
+/// So instead of using `type_of(const_arg)` we can use `type_of(const_param)` which does
+/// not depend on `typeck`.
```

Comment:
`typeck` of what?

Reply:
on any `typeck`. `type_of(const_param)` is fairly trivial as it is directly mentioned by the user.

We have `const N: usize` and the `type_of(N)` should return `usize`

Reply:
I cannot make sense of this sentence. Above, you use things like `typeck(main)`.

You seem to assume that the reader knows the half a dozen queries involved in this and how they related, and maybe that's fair, but I know none of them.^^